File: backend/app/services/split.py
from sqlalchemy.orm import Session
from .. import models

def run_monthly_financial_split(db: Session, mes_referencia: str, receita_bruta_adsense: float) -> dict:
    """
    Computes the monthly revenue distribution (Regra 70/30).
    Saves a record to FechamentoFinanceiro and returns the breakdown.
    """
    # 30% retention for management
    retencao_adm_30 = receita_bruta_adsense * 0.30
    # 70% share for eligible creators
    fundo_partilha_70 = receita_bruta_adsense * 0.70

    # Get active adimplente condôminos who are eligible (status must be ATIVO_ADIMPLENTE)
    eligible_condominos = db.query(models.Condomino).filter(
        models.Condomino.status == models.OperacionalStatus.ATIVO_ADIMPLENTE
    ).all()

    count_eligible = len(eligible_condominos)
    valor_por_condomino = fundo_partilha_70 / count_eligible if count_eligible > 0 else 0.0

    # Save closing to DB
    fechamento = models.FechamentoFinanceiro(
        mes_referencia=mes_referencia,
        receita_bruta_adsense=receita_bruta_adsense,
        retencao_adm_30=retencao_adm_30,
        fundo_partilha_70=fundo_partilha_70,
        valor_por_condomino=valor_por_condomino,
        qtd_condominos_ativos=count_eligible
    )
    db.add(fechamento)
    db.commit()
    db.refresh(fechamento)

    return {
        "id": fechamento.id,
        "mes_referencia": fechamento.mes_referencia,
        "receita_bruta_adsense": fechamento.receita_bruta_adsense,
        "retencao_adm_30": fechamento.retencao_adm_30,
        "fundo_partilha_70": fechamento.fundo_partilha_70,
        "valor_por_condomino": fechamento.valor_por_condomino,
        "qtd_condominos_ativos": fechamento.qtd_condominos_ativos,
        "data_fechamento": fechamento.data_fechamento.isoformat(),
        "eligible_list": [c.nome_comercial for c in eligible_condominos]
    }

File: src/lib/services/split.ts
import { supabase } from "../supabase";

export async function runMonthlyFinancialSplit(
  mesReferencia: string,
  receitaBrutaAdsense: number
): Promise<any> {
  const retencaoAdm30 = receitaBrutaAdsense * 0.30;
  const fundoPartilha70 = receitaBrutaAdsense * 0.70;

  // Filter condominos by ATIVO_ADIMPLENTE
  const { data: eligibleCondominos, error: fetchError } = await supabase
    .from("condominos")
    .select("*")
    .eq("status", "ATIVO_ADIMPLENTE");

  if (fetchError || !eligibleCondominos) {
    throw new Error(`Failed to fetch eligible condominos for split: ${fetchError?.message}`);
  }

  const countEligible = eligibleCondominos.length;
  const valorPorCondomino = countEligible > 0 ? fundoPartilha70 / countEligible : 0.0;

  // Save closing to Supabase
  const { data: fechamento, error: insertError } = await supabase
    .from("fechamentos_financeiros")
    .insert({
      mes_referencia: mesReferencia,
      receita_bruta_adsense: receitaBrutaAdsense,
      retencao_adm_30: retencaoAdm30,
      fundo_partilha_70: fundoPartilha70,
      valor_por_condomino: valorPorCondomino,
      qtd_condominos_ativos: countEligible,
    })
    .select()
    .single();

  if (insertError || !fechamento) {
    throw new Error(`Failed to save fechamento financeiro: ${insertError?.message}`);
  }

  return {
    id: fechamento.id,
    mes_referencia: fechamento.mes_referencia,
    receita_bruta_adsense: fechamento.receita_bruta_adsense,
    retencao_adm_30: fechamento.retencao_adm_30,
    fundo_partilha_70: fechamento.fundo_partilha_70,
    valor_por_condomino: fechamento.valor_por_condomino,
    qtd_condominos_ativos: fechamento.qtd_condominos_ativos,
    data_fechamento: fechamento.data_fechamento,
    eligible_list: eligibleCondominos.map((c) => c.nome_comercial),
  };
}
